### PR TITLE
111 running long tasks on mmoda

### PR DESCRIPTION
**Pull request by okolo**
passing callback to oda_api through .oda_api_callback file was implemented

### CODE REVIEW

**Review by volodymyrss — APPROVED** [review]

**Comment by volodymyrss** [comment]
Looks ok to me! oda_api still needs to be updated of course.

File: tests/test_callback.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        logger.info('service returns %s %s', r, r.json)
+
+        if r.json['workflow_status'] == 'done':
+            logger.info('workflow done!')
+            break
+
+        time.sleep(0.1)
+
+    test_worker_thread.join()
+    assert r.json['data']['output']['callback'] == callback_url
```
Comment (dsavchenko) [inline tests/test_callback.py:56]: Also need to assert that callback.json file is properly populated

File: nb2workflow/nbadapter.py [hunk 2]
```diff
@@ -407,12 +407,13 @@ def update_summary(self, **d):
 
         
 
-    def execute(self, parameters, progress_bar = True, log_output = True, inplace=False):
+    def execute(self, parameters, progress_bar = True, log_output = True, inplace=False, callback=None):
```
Comment (dsavchenko) [inline nb2workflow/nbadapter.py:410]: Parameter name `callback` is a bit misleading, `callback_url` is probably better
The same applies to `_pass_callback()` method.

**Comment by volodymyrss** [comment]
@okolo , I vaguely agree with comments of @dsavchenko . Please see to consider them and respond.

File: nb2workflow/nbadapter.py (hunk 2 above)
Comment (burnout87) [reply nb2workflow/nbadapter.py:410]: I guess also for the `_execute` method?

File: nb2workflow/nbadapter.py [hunk 3]
```diff
@@ -487,6 +490,17 @@ def _execute(self, parameters, progress_bar = True, log_output = True, inplace=F
 
         return exceptions
 
+    def _pass_callback(self, workdir: str, callback: str):
+        """
+        save callback to file .oda_api_callback in the notebook dir were it can be accessed by ODA API
```
[the remark this answers is not in the document]
Comment (burnout87) [reply nb2workflow/nbadapter.py:495]: As mentioned by @dsavchenko in a previous comment, it's better to change the name of the arg `callback` into `callback_url`

**Review by burnout87 — COMMENTED** [review]
In line with the points raised by @dsavchenko

File: tests/test_callback.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+@pytest.fixture
+def app():
+    app = nb2workflow.service.app
+    app.notebook_adapters = nb2workflow.nbadapter.find_notebooks('tests/testfiles')
+    nb2workflow.service.setup_routes(app)
+    print("creating app")
+    return app
+
+
+def test_progress_callback(client):
```
[the remark this answers is not in the document]
Comment (okolo) [reply tests/test_callback.py:18]: Here the idea is to access the callback URL from the test notebook tests/testfiles/callback.ipynb which is being executed asynchronously and return it as a result. In this way we check that the callback URL is passed properly.

File: tests/test_callback.py (hunk 1 above)
Comment (okolo) [reply tests/test_callback.py:56]: In the particular test the idea was to access the callback URL from the test notebook tests/testfiles/callback.ipynb which is being executed asynchronously and return it as a result. In this way we check that the callback URL is passed properly. The content of the callback.json has no connection to this particular test.

File: nb2workflow/nbadapter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (okolo) [reply nb2workflow/nbadapter.py]: I have made refactoring

**Review by burnout87 — APPROVED** [review]

File: tests/test_callback.py (hunk 1 above)
Comment (volodymyrss) [reply tests/test_callback.py:56]: We can add to the notebook also a `report_progress` and check that at the end, callbacks for `progress` and for `done` are stored.

Need to be sure callbacks are not overwriting, but appending to the file.

Comment (okolo) [reply tests/test_callback.py:59]: @volodymyrss   the test you were asking for was added in [rev. bb2ca0901b435af0700ea7f992959550f9426278 ](https://github.com/oda-hub/nb2workflow/pull/121/commits/bb2ca0901b435af0700ea7f992959550f9426278)

**Comment by dsavchenko** [comment]
I would like to approve, but the CI action fails. 
There is a problem installing owlready2 there. Its version is fixed to a really old one in requirements. @volodymyrss was there a particular reason for this?

**Comment by volodymyrss** [comment]
> I would like to approve, but the CI action fails. There is a problem installing owlready2 there. Its version is fixed to a really old one in requirements. @volodymyrss was there a particular reason for this?

I suppose there was a reason. Can you try to release the constrain, @okolo ?

**Review by dsavchenko — APPROVED** [review]

**Pull request merged**